### PR TITLE
issue #710 : add event list and or operator

### LINKED ISSUE #710: Handler Events as flags
Idea is to provide the following flexibility:
```python

@trainer.on(Events.ITERATION_STARTED(once=1) | Events.ITERATION_STARTED(every=10) | Events.COMPLETED)
def execute_some_handler(_):
    pass
```
such that `execute_some_handler` is triggered at the first iteration and at 10th, 20th, ...etc and in the end of the training.

### PR DESCRIPTION
**Pull request by sdesrozis**
Fixes #710  

Description:

Add an event list class to pack events which could be added with operator `|` on `Events` and `CallableEventWithFilter`

User code is
```
    def fn(e, b):
        pass

    trainer = Engine(fn)

    @trainer.on(Events.ITERATION_STARTED(once=1) | Events.ITERATION_STARTED(every=3) | Events.COMPLETED)
    def execute_some_handler(engine):
        print(engine.state.iteration)

    trainer.run(range(3), max_epochs=5)
```
TODO : improve doc

Check list:
* [x] New tests are added (if a new feature is added)
* [ ] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/engine/events.py [hunk 1]
```diff
@@ -179,6 +182,33 @@ def call_once(engine):
...
+
+
+class EventList:
+
+    def __init__(self):
+        self._events = []
+
+    def append(self, event: Union[Events, CallableEventWithFilter]):
+        if not isinstance(event, (Events, CallableEventWithFilter)):
+            raise ValueError(f"argument event should be Events or CallableEventWithFilter, got: {type(event)}")
```
Comment (vfdev-5) [inline ignite/engine/events.py]: This does not work in Py35 :(

File: tests/ignite/engine/test_custom_events.py [hunk 2]
```diff
@@ -433,3 +433,36 @@ def test_distrib_gpu(distributed_context_single_node_nccl):
...
+
+    iterations = []
+
+    @engine.on(Events.ITERATION_STARTED(once=1) | Events.ITERATION_STARTED(every=3) | Events.COMPLETED)
+    def execute_some_handler(e):
+        iterations.append(e.state.iteration)
+
+    engine.run(range(3), max_epochs=5)
+
+    assert iterations == [1, 3, 6, 9, 12, 15, 15]
```
Comment (vfdev-5) [inline tests/ignite/engine/test_custom_events.py]: Can we check that with mock and number of calls

File: ignite/engine/events.py [hunk 3]
```diff
@@ -179,6 +182,33 @@ def call_once(engine):
     GET_BATCH_STARTED = "get_batch_started"
     GET_BATCH_COMPLETED = "get_batch_completed"
 
+    def __or__(self, other):
+        return EventList() | self | other
+
+
+class EventList:
```
Comment (vfdev-5) [inline ignite/engine/events.py]: Shouldn't we inherit of something instead of introducing new type of events ?

Comment (sdesrozis) [reply ignite/engine/events.py]: I didn't introduce a new type of event but a list of event with operator `__or__`.

File: ignite/engine/events.py (hunk 1 above)
Comment (sdesrozis) [reply ignite/engine/events.py]: Ouch you're right, I remove `fstring`.

File: ignite/engine/engine.py [hunk 4]
```diff
@@ -334,7 +334,11 @@ def on(self, event_name, *args, **kwargs):
         """
 
         def decorator(f: Callable) -> Callable:
-            self.add_event_handler(event_name, f, *args, **kwargs)
+            if type(event_name) is EventList:
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: Seems like EventList are only supported for decorator. It remains `engine.add_event_handler`

Comment (sdesrozis) [reply ignite/engine/engine.py]: Yep, I didn't realize that. I fix it.

File: ignite/engine/engine.py [hunk 5]
```diff
@@ -244,6 +244,10 @@ def print_epoch(engine):
             See :class:`~ignite.engine.Events` for more details.
 
         """
+        if isinstance(event_name, EventList):
+            for e in event_name:
+                self.add_event_handler(e, handler, *args, **kwargs)
+            return
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: Shouldn't we still return something here ?
Otherwise, the docstring is incomplete.

Comment (sdesrozis) [reply ignite/engine/engine.py]: Indeed, we should return something, I didn't catch that. Taking a closer look, I think `add_event_handler` shouldn't be the way to add a list of events. A specific function should be defined.
@vfdev-5 thoughts ?

Comment (sdesrozis) [reply ignite/engine/engine.py]: `add_eventlist_handler` which returns a list of `RemovableEventHandler` ?

Comment (sdesrozis) [reply ignite/engine/engine.py]: See new commit. Skeptical about `add_eventlist_handler` naming... `add_event_handlers` ?

**Comment by vfdev-5** [comment]
@sdesrozis I'm not a fan of introducing another public method for that.
Idea of returning `RemovableEventHandler` is that it can be used as a context manager to remove attached handlers on exit of the context manager. I think we need to keep single method `add_event_handler` and return something such that it could be used as `RemovableEventHandler`... What do you think ?

**Comment by sdesrozis** [comment]
Now `add_event_handler` returns `RemovableEventHandler` compatible with `EventsList`.

TODO doc

**Comment by vfdev-5** [comment]
Sorry, my bad, the name should remain as previously as `RemovableEventHandle` and not `RemovableEventHandLER`.
Yes, now, it looks good for me. Please, add a test with `RemovableEventHandle` for event list.

**Comment by sdesrozis** [comment]
Last missing point is 
* README.md in "Power of Events & Handlers"

**Comment by sdesrozis** [comment]
@vfdev-5 you can comment :)

**Comment by vfdev-5** [comment]
@sdesrozis looks good to me. For perfect fit, I would put added section just after "Built-in events filtering"...

**Comment by vfdev-5** [comment]
Let me fix coverage relative decrease: https://codecov.io/gh/pytorch/ignite/pull/868/diff#diff-aWduaXRlL2VuZ2luZS9ldmVudHMucHk=
- https://codecov.io/gh/pytorch/ignite/pull/868/diff#D6-218
- https://codecov.io/gh/pytorch/ignite/pull/868/diff#D6-336

**Pull request merged**